optuna/storages/base.py

```diff
@@ -24,162 +24,848 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
 
     Storage classes abstract a backend database and provide library internal interfaces to
     read/write history of studies and trials.
+
+    Storage class can assume that the returned values are read-only. Additionally, storage
+    classes can assume that a single process has at most one storage. However, the storage
+    might be shared from multiple threads and storage classes must be thread-safe.
+
+    Storage classes must support monotonic-reads consistency model, that is, if a
+    process reads a data `X`, any successive reads on data `X` does not return
+    older values.
+    They must virtually support read-your-writes, that is, if a process writes to
+    data `X`, any successive reads on data `X` from the same process must read
+    the written value or one of more recent values.
+    If `sync` method is called, any successive reads on `state` attribute
+    of `Trial` and any attributes of `Study` must be equal or more recent than the value
+    at the time that the `sync` method is called.
+    If `sync` method is called, any successive reads on any attributes of `Trial` `T`
+    must return the same or more recent value than any preceding writes to `T` via
+    `update_trial` method.
+    Writes on any attributes of `Trial` must be completed before any
+    successive writes on the `state` attribute of the `Trial` with the same trial ID.
+
```

Comment:
It might be better to explicitly require that any finished trials should not be updated. Otherwise, storage classes must read all trials on every `sync` calls.